main.cpp

```diff
@@ -125,31 +188,62 @@ void uthread_yield(){
     if (ret_val == 1) return;
 
     // Choose the next thread to execute.
-    int id = current_thread_id;
-    do {
-        id = (id + 1) % num_threads;
-        if(id == current_thread_id){
-            printf("infinite loop!\n");
-        }
-    } while (threads[id].complete);
+    ready_list.push_back(current_thread_id);
+
+    current_thread_id = ready_list.front();
+    ready_list.erase(ready_list.begin());
```

Comment:
Thrread_yield is context switch specific for threads that want to be immediately placed back on the ready list.  That's not sufficient for threads that are blocking or completing execution.